### PR TITLE
Support remoteUiUrl from the mobile_app integration.

### PR DESCRIPTION
**Pull request by JBassett**
Fixes: https://github.com/home-assistant/home-assistant-android/issues/92

### CODE REVIEW

**Comment by quthla** [comment]
What if the cloud subscription ended or the remote UI is unavailable?

**Comment by JBassett** [comment]
Currently the mobile_app component won't tell us if the subscription is expired.  I don't think currently there is a way to determine if the from the apps point of view.  For now you would need to clear data and sign back in.

As far as the unavailable goes I'm working on that still.

**Comment by quthla** [comment]
And the internal / external feature of the iOS app? It has 2 prefs for the URLs and one for the home wifi. If connected to home wifi, it'll use the internal url and if not, the external.

**Comment by balloob** [comment]
We can add the internal/external feature later. Right now if a cloud user is adding their Home Assistant installation (and especially once we add discovery), it won't work when they leave their house.

File: data/src/main/java/io/homeassistant/companion/android/data/integration/IntegrationRepositoryImpl.kt [hunk 1]
```diff
@@ -102,6 +103,11 @@ class IntegrationRepositoryImpl @Inject constructor(
         return localStorage.getString(PREF_WEBHOOK_ID) != null
     }
 
+    override suspend fun getUiUrl(isInternal: Boolean): URL? {
+        return localStorage.getString(PREF_REMOTE_UI_URL)?.toHttpUrl()?.toUrl()
```
Comment (balloob) [inline data/src/main/java/io/homeassistant/companion/android/data/integration/IntegrationRepositoryImpl.kt]: We can use `isInternal` here already, by making sure we only return `authenticationRepository.getUrl()` if it's true.

**Comment by David0331** [comment]
O

**Comment by rajput306** [comment]
thanks for this [information](http://www.kztrading.in/).

File: data/src/main/java/io/homeassistant/companion/android/data/integration/IntegrationRepositoryImpl.kt (hunk 1 above)
Comment (JBassett) [reply data/src/main/java/io/homeassistant/companion/android/data/integration/IntegrationRepositoryImpl.kt]: Done.

**Comment by JBassett** [comment]
After looking into the auto switching URLs on errors it felt outside the scope of this PR so I deferred to a later time.

**Comment by quthla** [comment]
What I've got in mind is something like this:

Prefs: Home URL, Home Wi-Fi, Remote URL

Discovery: fills Home and Remote URL (if remote_ui_url available)

If only one URL filled: use that
If both URLs filled but no Home Wi-Fi selected: use Remote URL
If all 3 prefs filled: use Home URL when connected to Home Wi-Fi, else use Remote URL
If URL can't be reached on app start: show pref screen with error message so user can make necessary changes

This way, no auto switching remote url on errors is needed and it is ensured that the app never becomes unusable

Your opinions on this?

**Comment by balloob** [comment]
Yeah agree. Just realized that we should also get auth tokens from remote url when not at home. It's not just the web UI.

**Comment by JBassett** [comment]
Agreed, I noticed the auth issue too, I've already started working on extracting the URL handling so it can be reused in all places.

**Pull request merged**